Add release notes API record feature

saschagrunert opened the pull request:
This patch adds the ability to record the release notes API via two new
release notes command line flags `--record <DIR>` and `--replay <DIR>`.

This allows us to create custom test cases and implement changes to the
release notes tool faster without actually having the need to interact
with the GitHub API.

How does it work? A recorded session can be created for example via:

```
> go run cmd/release-notes/main.go \
    --github-token $TOKEN \
    --branch release-1.16 \
    --start-rev v1.16.2 \
    --end-rev v1.16.3 \
    --output v1.16.3.md \
    --record rec
```

Now the created directory `rec` contains all the serialized API calls:

```
> ls -1 rec | head
GetCommit-0.json
GetCommit-1.json
GetPullRequest-0.json
GetPullRequest-1.json
GetPullRequest-2.json
GetPullRequest-3.json
GetPullRequest-4.json
GetPullRequest-5.json
GetPullRequest-6.json
GetPullRequest-7.json
```

To replay the session without having the need for a valid token, we can
now invoke:

```
> go run cmd/release-notes/main.go \
    --replay rec \
    --output v1.16.3-replayed.md
```

Signed-off-by: Sascha Grunert <sgrunert@suse.com>

saschagrunert commented:
Ready for review. This allows us to record the release notes API and replay for later (mocked) integration testing.

hoegaarden commented on pkg/notes/options/options.go:
```diff
@@ -0,0 +1,376 @@
+
+	"k8s.io/release/pkg/git"
+	"k8s.io/release/pkg/notes/client"
+	"k8s.io/release/pkg/notes/client/clientfakes"
```
I strongly opt against mixing test- and prod-code.

If I understand you proposed change correctly, all you'd need to do to make that happen is to implement two or more new clients. E.g. a `notes.RecordingClient` and a `notes.ReplayingClient`. Both would need to implement the [`Client` interface][clientInt]. When you set up the [`Gatherer`][gatherer], you can decide, based on your command line flag/options/... which client you hand in:
- a [wrapped github client][wrappedClient] (the "default" client)
- a `RecordingClient`
- a `ReplayingClient`

The `RecordingClient` could just wrap a real client `client := notes.RecordingClient{Client: client, Recordings: "/some/dir"}` and intercept & save the responses, the `ReplayingClient` just needs the dump dir `client := notes.ReplayingClient{Recordings: "/some/dir"}`. Those different clients could be nicely tested in isolation and we'd have a nice separation of concerns.

The options package and struct do way too much for my taste anyway, I don't think we should cram more functionality onto it.

Also, in general this seems like a test concern and I am not sure if we'd want to expose that via the `krel` (or any other) binary, but we should maybe see to make those test cases / create a test harness.  But if you think exposing that via the binaries is useful, I certainly won't block that!


[clientInt]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L31-L35
[gatherer]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/cmd/krel/cmd/changelog.go#L215-L220
[wrappedClient]: https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L41-L43

saschagrunert replied:
> I strongly opt against mixing test- and prod-code.
> 
> If I understand you proposed change correctly, all you'd need to do to make that happen is to implement two new clients. E.g. a `notes.RecordingClient` and a `notes.ReplayingClient`. Both would need to implement the [`Client` interface](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L31-L35). When you set up the [`Gatherer`](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/cmd/krel/cmd/changelog.go#L215-L220), you can decide, based on your command line flag/options/... which client you hand in:
> 
> * a [wrapped github client](https://github.com/kubernetes/release/blob/2c9ce919c182eb72801775c036861a29aa1e88e9/pkg/notes/client.go#L41-L43) (the "default" client)
> * a `RecordingClient`
> * a `ReplayingClient`
>
> The `RecordingClient` could just wrap a real client `client := notes.RecordingClient{Client: client, Recordings: "/some/dir"}` and intercept & save the responses, the `ReplayingClient` just needs the dump dir `client := notes.ReplayingClient{Recordings: "/some/dir"}`. Those different clients could be nicely tested in isolation and we'd have a nice separation of concerns.
> 
> The options package and struct do way too much for my taste anyway, I don't think we should cram more functionality onto it.

Alright, thank you for the input. I separated both implementations and putt them into a dedicated `recordreplay` package. 

> Also, in general this seems like a test concern and I am not sure if we'd want to expose that via the `krel` (or any other) binary, but we should maybe see to make those test cases / create a test harness. But if you think exposing that via the binaries is useful, I certainly won't block that!

Yeah I thought it would be not only good for testing, but also for development: If I want to change the layout/parsing of the release notes it would be good to be able to fastly replay the latest recording without having the need for any GitHub API access. WDYT?

hoegaarden commented on pkg/notes/options/options.go:
```diff
@@ -151,5 +178,63 @@ func (o *Options) ValidateAndFinish() error {
+	o.ReplayDir = previousReplayDir
+	o.gitCloneFn = git.CloneOrOpenGitHubRepo
+	return nil
+}
```
I think I understand why this (forcing the same options as were used in the recorded session) happens, but I think the least we should do then is to make that really clear to the user, in other words log a warning that basically all your options/flags will be ignored.

saschagrunert replied:
I will remove the options de-/serialization to avoid confusion here.

hoegaarden commented on cmd/krel/cmd/changelog.go:
```diff
@@ -205,22 +204,18 @@ func generateReleaseNotes(branch, startRev, endRev string) (string, error) {
-		Repo:    git.DefaultGithubRepo,
+		Client:  c,
+		Context: context.Background(),
+		Options: notesOptions,
```
I find it confusing that I pass the Client to the gatherer but also the options, which seem to be the canonical way on how to create a client. Why both?

saschagrunert replied:
I changed the gatherer to have only private fields but two convenience `New...` functions. This makes the overall usage more straight forward.

hoegaarden commented on pkg/notes/client/recordreplay/record_client.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+type githubNotesRecordClient struct {
+	client      *client.GithubNotesClient
```
I think that should be the interface, not the discrete type.

```suggestion
	client      *client.Client
```

saschagrunert replied:
Fixed :+1:

saschagrunert commented on pkg/notes/client/recordreplay/replay_client.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (c *githubNotesReplayClient) GetCommit(ctx context.Context, owner, repo, sha string) (*github.Commit, *github.Response, error) {
+	return c.client.GetCommit(ctx, owner, repo, sha)
```
Then I would have to track the state (amount of counts) in the same manner as the record client. Do you prefer that approach?

saschagrunert replied:
Removed the fake client from the commit.

saschagrunert commented on pkg/notes/client/recordreplay/replay_client.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+type githubNotesReplayClient struct {
+	client    *clientfakes.FakeClient
```
See discussion on: https://github.com/kubernetes/release/pull/1039#discussion_r372303726

saschagrunert commented on pkg/notes/client/client.go:
```diff
@@ -38,55 +38,55 @@ type Client interface {
 
-func WrapGithubClient(ghc *github.Client) Client {
-	return &githubNotesClient{ghc: ghc}
+func NewGithubNotesClient(ghc *github.Client) *GithubNotesClient {
```
Yes, this can be improved (incoming with the next commit). :+1:

saschagrunert replied:
Fixed that as well :)

saschagrunert commented on pkg/notes/options/options.go:
```diff
@@ -151,5 +178,63 @@ func (o *Options) ValidateAndFinish() error {
+
+	if o.RecordDir != "" {
+		return recordreplay.NewGithubNotesRecordClient(c, o.RecordDir), nil
+	}
```
Yes, I put everything into the `client` package.

saschagrunert commented on pkg/notes/client/recordreplay/record_client.go:
```diff
@@ -0,0 +1,150 @@
+	if j, ok := c.recordState[api]; ok {
+		i = j + 1
+	}
+	c.recordState[api] = i
```
I need the `i` later on for the filename :)

saschagrunert commented on pkg/notes/client/recordreplay/record_client.go:
```diff
@@ -0,0 +1,150 @@
+	api GitHubAPI, result interface{}, response *github.Response,
+) error {
+	if c.recordDir == "" {
+		return nil
```
Removed that check :+1:

saschagrunert commented:
> All in all I think this is a great feature!
> 
> > Yeah I thought it would be not only good for testing, but also for development: If I want to change the layout/parsing of the release notes it would be good to be able to fastly replay the latest recording without having the need for any GitHub API access. WDYT?
> 
> 
> 
> Given I understand your approach and proposed feature correctly, my remaining main concerns are:
> 
> * Nit: naming things an slice them into packages: things like `client.Client` and `options.Options` always read weird to me. I think `golint` would also flag them as "stutter".
> * The replay client could be way easier and should not mix prod- and test-code (the counterfeiter client), and frankly I can't see any reason why it would need to wrap any client, neither test/fake nor real. Am I missing something?
> * It feels a bit strange to me that we kinda pass 2 potentially different clients into the gatherer, one directly and one via the options. Is that just me maybe?

I changed the layout of the packages and addressed most of your review notes. Thanks for the detailed review :pray:  I think we're moving into the right direction. PTAL. :)

saschagrunert commented:
/test pull-release-cluster-up

saschagrunert commented on pkg/notes/client/record.go:
```diff
@@ -0,0 +1,158 @@
+	GitHubAPIListPullRequestsWithCommit GitHubAPI = "ListPullRequestsWithCommit"
+	GitHubAPIGetPullRequest             GitHubAPI = "GetPullRequest"
+	GitHubAPIGetRepoCommit              GitHubAPI = "GetRepoCommit"
+)
```
No, I will make them private.

saschagrunert commented on pkg/notes/client/clientfakes/fake_client.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2020 The Kubernetes Authors.
```
Ah cool, awesome! :+1:

saschagrunert commented on pkg/notes/options/options.go:
```diff
@@ -151,5 +172,34 @@ func (o *Options) ValidateAndFinish() error {
 }
+
+// Client returns either a fake client or a real GitHub API client based on if
+// the `StaticAPI` option is set or not
```
Thank for the input, I will change the comment as suggested.

saschagrunert commented on pkg/notes/client/client.go:
```diff
@@ -38,55 +38,55 @@ type Client interface {
 
-type githubNotesClient struct {
-	ghc *github.Client
+type GithubNotesClient struct {
```
No, I will make it private again.

hoegaarden commented:
/lgtm
/approve
/hold cancel
